openmmtools/multistate/multistatesampler.py

```diff
@@ -1774,10 +1775,16 @@ def _update_timing(self, iteration_time, partial_total_time, run_initial_iterati
     def _display_cuda_devices():
         """Query system nvidia-smi to get available GPUs indices and names in debug log."""
         # Read nvidia-smi query, should return empty strip if no GPU is found.
-        cuda_query_output = os.popen("nvidia-smi --query-gpu=index,gpu_name --format=csv,noheader").read().strip()
-        # Split by line jump and comma
-        cuda_devices_list = [entry.split(',') for entry in cuda_query_output.split('\n')]
-        logger.debug(f"CUDA devices available: {*cuda_devices_list,}")
+        cuda_query_output = subprocess.run("nvidia-smi --query-gpu=gpu_uuid,gpu_name,compute_mode  --format=csv", shell=True, capture_output=True, text=True)
+        print(cuda_query_output)
+        if cuda_query_output.stdout:
+            # Split by line jump and comma
+            cuda_devices_list = [entry for entry in cuda_query_output.stdout.splitlines()]
+            logger.debug(f"CUDA devices available: {*cuda_devices_list,}")
+            if "Default" in cuda_query_output.stdout:
+                logger.warn(f"GPU in 'Exclusive_Process' mode, one context is allowed per device. This may prevent some openmmtools features from working.")
+        else:
+            logger.debug(f"nvidia-smi command failed: {cuda_query_output.stderr}, this is expected if there is no GPU available")
```

Comment:
This is looking good! I think here we probably want to check the return code, we could do that just by checking the `cuda_query_output.returncode` value (if it's 0 it should be okay), or we could also probably add the `check=True` argument to the subprocess call, which would return a `CalledProcessError if it fails, we can try catching that error to report it in the log.

Reply:
You know I think you are right, I was assuming that if `nvidia-smi` was not there, nothing would be written to stdout, so the logic check would work fine, but if there is a case where something is written to stdout and stderr, then this wouldn't work. I thought this was cleaner than catching a `CalledProcessError`, but I think there is an edge case where it doesn't work. I will refactor it to catch the error (which is easy, just need to replace the if/esle with try/except, and run with check=True inside the try block. Oh and I need to remove the `print` there that was for debugging.